instr.ml

```diff
@@ -244,24 +262,70 @@ let used_vars = function
   | Comment _
   | Read _
   | Stop -> VarSet.empty
-  | Osr (e, _, _, osr) ->
+  | Osr (e, _, _, _, osr) ->
     let exps = List.map (function
         | OsrConst (_, e) -> e
         | OsrMut (_, x) -> Simple (Var x)
         | OsrMutUndef _ -> Simple (Lit Nil)) osr in
     let exps_vars = List.map expr_vars exps in
     List.fold_left VarSet.union (expr_vars e) exps_vars
 
-type 'a dict = (string * 'a) list
-
-type version = label * instruction_stream
-type program = version list
-
-let active_version (prog : program) : version =
-  (List.hd prog)
-
-let replace_active_version (prog : program) (repl : version) =
-  repl :: (List.tl prog)
+type scope_annotation =
+  | ExactScope of VarSet.t
+  | AtLeastScope of VarSet.t
+
+type inferred_scope =
+  | DeadScope
+  | Scope of ModedVarSet.t
+
+type annotations = scope_annotation option array
+
+type formal_parameter =
+  | ParamConst of variable
+  | ParamMut of variable
+
+type version = (label * instruction_stream)
+type version_annotation = (label * annotations)
+
+type afunction = {
+  name : label;
+  formals : formal_parameter list;
+  body : version list;
+  annotations : version_annotation list;
+}
+type program = {
+  main : afunction;
+  functions : afunction list;
+}
+
+exception FunctionDoesNotExist of label
+exception AmbiguousFunctionName of label
+
+let lookup_fun (prog : program) (name : label) : afunction =
+  if name = "main" then prog.main else
+  match List.filter (fun {name = l} -> name = l) prog.functions with
```

Comment:
```
try List.find (fun {name = l} -> name = l)
with Not_found -> raise (FunctionDoesNotExist name)
```

(it picks the first, but then do we really need to complain at this level? We could just ignore the issue or forbid name conflicts at declaration time.)